### PR TITLE
[IMP] payment, sale: prevent paying for a canceled SO/invoice

### PR DESCRIPTION
**Pull request by hote-odoo**
Before this commit, sales orders/invoices that had been canceled could
still be paid from a payment link. Now, when trying to pay for a
canceled sales order/invoice, the amount to pay is set to zero.
    
Task - 2735019

### CODE REVIEW

File: addons/sale/controllers/portal.py [hunk 1]
```diff
@@ -347,6 +349,8 @@ def _get_custom_rendering_context_values(self, sale_order_id=None, **kwargs):
         rendering_context_values = super()._get_custom_rendering_context_values(**kwargs)
         if sale_order_id:
             rendering_context_values['sale_order_id'] = sale_order_id
+        if sale_order_state:
+            rendering_context_values['sale_order_state'] = sale_order_state
```
Comment (Feyensv) [inline addons/sale/controllers/portal.py:353]:
```suggestion
            rendering_context_values['sale_order_id'] = sale_order_id
            order_sudo = request.env['sale.order'].sudo().browse(sale_order_id)
            if order_sudo.state == 'cancel':
                rendering_context_values['warning'] = _("Attention: this order has been cancelled.")
```

File: addons/sale/controllers/portal.py [hunk 2]
```diff
@@ -334,10 +334,12 @@ def payment_pay(self, *args, amount=None, sale_order_id=None, access_token=None,
                 'partner_id': order_sudo.partner_id.id,
                 'company_id': order_sudo.company_id.id,
                 'sale_order_id': sale_order_id,
+                'sale_order_state': order_sudo.state,
```
Comment (Feyensv) [inline addons/sale/controllers/portal.py:337]:
```suggestion
```

File: addons/payment/controllers/portal.py [hunk 3]
```diff
@@ -127,6 +127,9 @@ def payment_pay(
         # Generate a new access token in case the partner id or the currency id was updated
         access_token = payment_utils.generate_access_token(partner_sudo.id, amount, currency.id)
 
+        # The invoice status is required to warn the user/customer about cancelled invoices
+        invoice_state = request.env["account.move"].sudo().browse(invoice_id).state
+
```
Comment (Feyensv) [inline addons/payment/controllers/portal.py:132]: * Add a check before to verify that the invoice exists if provided (seems we forgot that in the payment apocalypse)
```python
        warning = ''
        if invoice_id:
            invoice_sudo = request.env['account.move'].browse(currency_id).exists()
           if not invoice_sudo:
                raise ValidationError(_("The provided parameters are invalid."))
           if invoice_sudo.state == 'cancel':
               warning = _("Attention: this invoice has been cancelled.")
```

File: addons/payment/controllers/portal.py [hunk 4]
```diff
@@ -142,6 +145,7 @@ def payment_pay(
             'res_company': company,  # Display the correct logo in a multi-company environment
             'partner_is_different': partner_is_different,
             'invoice_id': invoice_id,
+            'invoice_state': invoice_state,
```
Comment (Feyensv) [inline addons/payment/controllers/portal.py:148]:
```suggestion
            'warning': warning,
```

**Review by Feyensv — CHANGES_REQUESTED** [review]
I think it would be more clean and reusable to provide a generic `warning` mechanism on the payment template instead of giving `sale_order_state` & `invoice_state`

File: addons/payment/views/payment_templates.xml [hunk 5]
```diff
@@ -43,6 +43,14 @@
             <t t-set="footer_template_id"
                t-value="footer_template_id or 'payment.footer'"/>
 
+            <t t-if="invoice_state=='cancel'">
+                <div class="alert alert-danger" role="alert">
+                    <strong>
+                        Attention: this invoice has been cancelled.
+                    </strong>
+                </div>
+            </t>
```
[the remark this answers is not in the document]
Comment (hote-odoo) [reply addons/payment/views/payment_templates.xml]: done :)

File: addons/sale/controllers/portal.py (hunk 1 above)
Comment (hote-odoo) [reply addons/sale/controllers/portal.py]: done :)

File: addons/payment/controllers/portal.py (hunk 4 above)
Comment (hote-odoo) [reply addons/payment/controllers/portal.py]: done :)

File: addons/sale/controllers/portal.py (hunk 2 above)
Comment (hote-odoo) [reply addons/sale/controllers/portal.py]: done :)

File: addons/sale/views/payment_templates.xml [hunk 6]
```diff
@@ -6,6 +6,15 @@
...
         </xpath>
+        <xpath expr="//form[@name='o_payment_checkout']" position="before">
+            <t t-if="sale_order_state == 'cancel'">
+                <div class="alert alert-danger" role="alert">
+                    <strong>
+                        Attention: this order has been cancelled.
+                    </strong>
+                </div>
+            </t>
+        </xpath>
```
[the remark this answers is not in the document]
Comment (hote-odoo) [reply addons/sale/views/payment_templates.xml]: done :)

File: addons/payment/controllers/portal.py [hunk 7]
```diff
@@ -127,13 +127,23 @@ def payment_pay(
         # Generate a new access token in case the partner id or the currency id was updated
         access_token = payment_utils.generate_access_token(partner_sudo.id, amount, currency.id)
 
+        # First verify the existence of the invoice and second set the amount to zero in case
+        # of a cancelled invoice.
```
Comment (chevalierv) [inline addons/payment/controllers/portal.py:131]: Prefer to explain the why than to describe your code. Here is an example of what your comment could be:
```suggestion
        # If the invoice is cancelled, the payment link still allows to go through the payment. 
        # This code ensures that no payment can be made when the state of the invoice is cancelled.
```

File: addons/sale/controllers/portal.py [hunk 8]
```diff
@@ -347,6 +347,10 @@ def _get_custom_rendering_context_values(self, sale_order_id=None, **kwargs):
         rendering_context_values = super()._get_custom_rendering_context_values(**kwargs)
         if sale_order_id:
             rendering_context_values['sale_order_id'] = sale_order_id
+            order_sudo = request.env['sale.order'].sudo().browse(sale_order_id)
+            # I the SO is cancelled do not let the customer to pay any amount.
```
Comment (chevalierv) [inline addons/sale/controllers/portal.py:351]: Your missing an 'F' here :)

**Review by chevalierv — APPROVED** [review]
LGTM :+1: 
I've made two nitpicking comments :p

File: addons/sale/controllers/portal.py (hunk 8 above)
Comment (hote-odoo) [reply addons/sale/controllers/portal.py]: done :)

File: addons/payment/controllers/portal.py (hunk 7 above)
Comment (hote-odoo) [reply addons/payment/controllers/portal.py]: done :)

**Review by Feyensv — COMMENTED** [review]
LGTM :+1: 

PR & commit title/message should be updated to specify `account` & `sale` as modules, and `SO + invoice` ;)

**Comment by Feyensv** [comment]
> Before sale orders/invoices that has been cancelled could still
> be paid, from a payment link for example.
> No when prompted to pay for a cancelled sale order/invoice the
> amount will be set to zero.

Previously/Before this commit, sale orders/invoices that **have** been cancelled ...

**Now**, ...

;)

**Review by Feyensv — APPROVED** [review]

**Review by AntoineVDV — APPROVED** [review]
@robodoo r+

**Pull request closed without merging**